go/engine/identify2_with_uid.go

```diff
@@ -0,0 +1,457 @@
+	return &Identify2WithUID{
+		Contextified: libkb.NewContextified(g),
+		arg:          arg,
+	}
+}
+
+// GetPrereqs returns the engine prereqs.
+func (e *Identify2WithUID) Prereqs() Prereqs {
+	return Prereqs{}
+}
+
+// Run then engine
+func (e *Identify2WithUID) Run(ctx *Context) (err error) {
+
+	e.G().Log.Debug("+ Identify2WithUID.Run(UID=%v, Assertion=%s)", e.arg.Uid, e.arg.UserAssertion)
+	ch := make(chan error)
+	e.resultCh = ch
+	go e.run(ctx)
+	err = <-ch
+	e.G().Log.Debug("- Identify2WithUID.Run() -> %v", err)
+	return err
+}
+
+func (e *Identify2WithUID) run(ctx *Context) {
+	err := e.runReturnError(ctx)
+	e.unblock(err)
+}
+
+func (e *Identify2WithUID) runReturnError(ctx *Context) (err error) {
+
+	e.G().Log.Debug("+ acquire singleflight lock")
+	lock := locktab.AcquireOnName(e.arg.Uid.String())
+	e.G().Log.Debug("- acquired")
+
+	defer lock.Release()
+
+	if e.loadAssertion(); err != nil {
+		return err
+	}
+
+	if !e.useAnyAssertions() && e.checkFastCacheHit() {
+		e.G().Log.Debug("| hit fast cache")
+		return nil
+	}
+
+	e.G().Log.Debug("| Identify2WithUID.loadUsers")
+	if err = e.loadUsers(ctx); err != nil {
+		return err
+	}
+
+	if err = e.checkLocalAssertions(); err != nil {
+		return err
+	}
+
+	if e.isSelfLoad() {
+		e.G().Log.Debug("| was a self load")
+		return nil
+	}
+
+	if !e.useRemoteAssertions() && e.checkSlowCacheHit() {
+		e.G().Log.Debug("| hit slow cache, first check")
+		return nil
+	}
+
+	e.G().Log.Debug("| Identify2WithUID.createIdentifyState")
+	if err = e.createIdentifyState(); err != nil {
+		return err
+	}
+
+	if err = e.runIdentifyPrecomputation(); err != nil {
+		return err
+	}
+
+	// First we check that all remote assertions as present for the user,
+	// whether or not the remote check actually suceeds (hence the
+	// ProofState_NONE check).
+	okStates := []keybase1.ProofState{keybase1.ProofState_NONE, keybase1.ProofState_OK}
+	if err = e.checkRemoteAssertions(okStates); err != nil {
+		e.G().Log.Debug("| Early fail due to missing remote assertions")
+		return err
+	}
+
+	if e.useRemoteAssertions() && e.checkSlowCacheHit() {
+		e.G().Log.Debug("| hit slow cache, second check")
+		return nil
+	}
+
+	// If we're not using tracking and we're not using remote assertions,
+	// we can unblock the RPC caller here, and perform the identifyUI operations
+	// in the background.
+	if !e.useTracking && !e.useRemoteAssertions() {
+		e.unblock(nil)
+	}
+
+	if err = e.runIdentifyUI(ctx); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *Identify2WithUID) getNow() time.Time {
+	if e.testArgs != nil && e.testArgs.clock != nil {
+		return e.testArgs.clock()
+	}
+	return time.Now()
+}
+
+func (e *Identify2WithUID) unblock(err error) {
+	e.chMutex.Lock()
+	defer e.chMutex.Unlock()
+	e.G().Log.Debug("| unblocking...")
+	if e.resultCh != nil {
+		e.G().Log.Debug("| ...ready to go -> %v", err)
+		e.resultCh <- err
+		e.resultCh = nil
+	}
+}
```

Comment:
So you want to be able to call unblock multiple times and just use the first one that comes in?

Wouldn't a buffered resultCh accomplish this and obviate the need for the chMutex and setting a channel to nil?